nitrite_test.go

```diff
@@ -7,79 +7,141 @@ import (
+			debugAttestationB64,
+		},
+	}
+	for _, tt := range happyPathTests {
+		t.Run(tt.name, func(t *testing.T) {
+			// when
+			result, err := nitrite.Verify(
+				attestation,
+				nitrite.WithRootCert(roots),
+				nitrite.WithAttestationTime(),
+			)
 
-		result, err := nitrite.Verify(attestBytes, opts)
-		require.NoError(t, err)
-		assert.Equal(t, *result.Document, *doc)
-		assert.True(t, result.SignatureOK)
-	})
+			// then
+			require.NoError(t, err)
+			require.NotEmpty(t, result)
+		})
+	}
 
-	t.Run("certificate has expired or is not yet valid", func(t *testing.T) {
-		opts := nitrite.VerifyOptions{
-			CurrentTime: time.UnixMilli(0),
-			Roots:       roots,
-		}
+	unknownSigningAuthorityTests := []struct {
+		name        string
+		rootCertOpt nitrite.RootCertFunc
+	}{
+		{
+			"nil roots",
+			nitrite.WithRootCert(nil),
+		},
+		{
+			"empty roots",
+			nitrite.WithRootCert(x509.NewCertPool()),
+		},
+	}
+	for _, tt := range unknownSigningAuthorityTests {
+		t.Run(tt.name, func(t *testing.T) {
+			// when
+			_, err := nitrite.Verify(
+				attestation,
+				tt.rootCertOpt,
+				nitrite.WithAttestationTime(),
+			)
 
-		result, err := nitrite.Verify(attestBytes, opts)
-		assert.ErrorContains(t, err, "certificate has expired or is not yet valid")
-		assert.Nil(t, result)
-	})
+			// then
+			assert.ErrorContains(t, err, "verifying certificate")
+		})
+	}
+
+	timeOutOfBoundsTests := []struct {
+		name        string
+		timeOpt     nitrite.VerificationTimeFunc
+		errContains string
+	}{
+		{
+			"certificate expired",
+			nitrite.WithTime(attestationTime.Add(-time.Hour)),
+			"verifying certificate",
+		},
+		{
+			"certificate not yet valid",
+			nitrite.WithTime(attestationTime.Add(time.Hour * 8760)),
+			"verifying certificate",
+		},
+		{
+			"zero time",
+			nitrite.WithTime(time.Time{}),
+			"verification time is 0",
+		},
+	}
+	for _, tt := range timeOutOfBoundsTests {
+		t.Run(tt.name, func(t *testing.T) {
+			// when
+			_, err := nitrite.Verify(
+				attestation,
+				nitrite.WithDefaultRootCert(),
+				tt.timeOpt,
+			)
+
+			// then
+			assert.ErrorContains(t, err, tt.errContains)
+		})
+	}
 }
 
-func TestNitriteDocument_NewDocumentFromCosePayloadBytes(t *testing.T) {
-	attestBytes, err := base64.StdEncoding.DecodeString(nitroAttestString)
+func TestDocument_CreatedAt(t *testing.T) {
+	attestation, err := base64.StdEncoding.DecodeString(attestationB64)
 	require.NoError(t, err)
 
 	t.Run("happy path", func(t *testing.T) {
-		doc, err := nitrite.NewDocumentFromCosePayloadBytes(attestBytes)
-		require.NoError(t, err)
+		// given
+		wantTime := attestationTime
 
-		assert.Equal(t, doc.ModuleID, "i-0eec57d9c38705a57-enc018f25eda1b897b7")
-		assert.Equal(t, doc.Timestamp, uint64(1715103786613))
-		assert.Equal(t, doc.Digest, "SHA384")
-		assert.Empty(t, doc.UserData)
-		assert.Empty(t, doc.Nonce)
-	})
+		result, err := nitrite.Verify(
+			attestation,
+			nitrite.WithDefaultRootCert(),
+			nitrite.WithAttestationTime(),
+		)
```

Comment:
Add error check before using result to get document created at time
```golang
result, err := nitrite.Verify(
	attestation,
	nitrite.WithDefaultRootCert(),
	nitrite.WithAttestationTime(),
)
require.NoError(t, err)
```

Reply:
Actually, couldn't we simplify some of the setup by simply creating a `Document` instead of going through the whole Verify process.
```golang
func TestDocument_CreatedAt(t *testing.T) {
	t.Run("happy path", func(t *testing.T) {
		// given
		doc := nitrite.Document{Timestamp: uint64(attestationTime.UnixMilli())}
		wantTime := attestationTime

		// when
		gotTime := doc.CreatedAt()

		// then
		assert.Equal(t, wantTime.UTC(), gotTime.UTC())
	})

        // I added a test for zero time to make sure we get a time.Time{} struct back
	t.Run("zero time", func(t *testing.T) {
		// given
		wantTime := time.Time{}
		doc := nitrite.Document{Timestamp: uint64(0)}

		// when
		gotTime := doc.CreatedAt()

		// then
		assert.Equal(t, wantTime, gotTime)
	})
}
```

Reply:
Updated